ocaml/quicktest/quicktest_vdi_copy_data_integrity.ml

```diff
@@ -0,0 +1,188 @@
+         assert (Unix.LargeFile.lseek fd offset Unix.SEEK_SET = offset);
+         Unix.write fd buf 0 length |> ignore
+       in
+
+       let write_random_extent () =
+         let offset = Random.int64 size in
+         let to_end = Int64.sub size offset in
+         let max_extent_length =
+           min (Int64.of_int max_write_length) to_end |> Int64.to_int
+         in
+         let length = Random.int max_extent_length in
+
+         write_random_block offset length
+       in
+
+       let num_extents = Random.int max_writes in
+       for i = 1 to num_extents do
+         write_random_extent ()
+       done
+    )
+
+let fill rpc session_id vdi =
+  let size =
+    Client.Client.VDI.get_virtual_size ~rpc ~session_id ~self:vdi
+    |> Int64.to_int
+  in
+  with_open_vdi rpc session_id vdi `RW
+    (fun fd ->
+       let buf = random_bytes size in
+       Unix.write fd buf 0 size
+    )
+
+let checksum rpc session_id vdi =
+  with_attached_vdi rpc session_id vdi `RO (fun path ->
+      Digest.to_hex (Digest.file path)
+    )
+
+let check_vdi_unchanged test rpc session_id sR ~prepare_vdi =
+  with_test test (fun () ->
+      let vdi = Client.Client.VDI.create ~rpc ~session_id ~name_label:"" ~name_description:"" ~sR ~virtual_size:4194304L ~_type:`user ~sharable:false ~read_only:false ~other_config:[] ~xenstore_data:[] ~sm_config:[] ~tags:[] in
+      finally
+        (fun () ->
+           prepare_vdi vdi;
+           let checksum_original = checksum rpc session_id vdi in
+           let copy = Client.Client.VDI.copy ~rpc ~session_id
+               ~vdi
+               ~base_vdi:API.Ref.null
+               ~into_vdi:API.Ref.null
+               ~sr:sR
+           in
+           finally
+             (fun () ->
+                let checksum_copy = checksum rpc session_id vdi in
+                if (checksum_copy <> checksum_original) then
+                  failwith (Printf.sprintf "VDI copy (checksum: %s) has different data than original (checksum: %s)." checksum_copy checksum_original)
+             )
+             (fun () -> Client.Client.VDI.destroy ~rpc ~session_id ~self:copy)
+        )
+        (fun () -> Client.Client.VDI.destroy ~rpc ~session_id ~self:vdi)
+    )
+
+let test_sr rpc session_id sr =
+  let sr_name = Client.Client.SR.get_name_label ~rpc ~session_id ~self:sr in
+  let test = Quicktest_common.make_test ("VDI.copy on SR [" ^ sr_name ^ "]") 4 in
+  with_test test (fun () ->
+      let test_nodata = Quicktest_common.make_test "Copy of empty VDI" 6 in
+      check_vdi_unchanged test_nodata rpc session_id sr
+        ~prepare_vdi:(fun _vdi -> ());
+
+      let test_random = Quicktest_common.make_test "Copy of random VDI" 6 in
+      check_vdi_unchanged test_random rpc session_id sr
+        ~prepare_vdi:(write_random_data rpc session_id);
+
+      let test_full = Quicktest_common.make_test "Copy of full VDI" 6 in
+      check_vdi_unchanged test_full rpc session_id sr
+        ~prepare_vdi:(fill rpc session_id)
```

Comment:
This is a good start. Might be useful to explicitly cover a few edge cases: hole at beginning of file (multiple of 64k), hole at end of file (multiple of 64k), write some data, then write random multiple of 64k zeroes (don't know if qemu would convert the zeroes to sparse, it might).

Also how do snapshots and leaf-coalesce affect this functionality? Might be useful to add a few tests that take write some random data with holes, take a snapshot, modify the original vdi again, check export of original vdi/snapshotted vdi, then delete snapshot, and do some copies again while SMGC is running. Although if there are any bugs there they would be in qemu/SMGC, so perhaps such a test is better written as part of the pool stress test. Could you verify if there are any such tests in the pool stress, and if not create a ticket to add some?

Reply:
Why should we write multiples of 64K blocks in the test? In my opinion completely random block lengths / offsets cover more edge cases. I could add another test case though for writing zeroes into the VDI.
I agree that those more complex scenarios should be run by XenRT. I'll check what VDI.copy tests we have at the moment.

Reply:
Right, I think that you need a minimum of 64k, not necesarily a multiple of 64k. So 64k+random would cover more edge cases true.

Reply:
Not sure how we could detect this from a quicktest, maybe by checking the `physical_utilization` field if that's reliable. But there is a XenRT test that was failing because the copy did not took advantage of sparseness.

Reply:
The VDI size is 4M, hence there is a reasonable chance that we will have >64k in our test. More tests could be added in a separate PR.